source/SH3/graphics/texture.cpp

```diff
@@ -9,52 +9,326 @@
+
+                ++y;
+                if(y % 2 == 0)
+                {
+                    // each iteration we read two lines: offset 0 and 2
+                    // so after two iterations, we have read the lines with offset 0 and 2, 1 and 3
+                    // so now we need to skip two additional lines, since we read 2 and 3 already
+                    y += 2;
+                    ASSERT(y <= header.texHeight);
+                    offsetFlipper = !offsetFlipper;
+                }
+
+                if(y == header.texHeight)
+                {
+                    break;
+                }
+            }
+
+            // Christ help me...
+            for(std::size_t pindex = 0, doff = 0; pindex < iBuffer.size() && doff < data.size(); pindex++, doff += 3)
+            {
+                ASSERT(iBuffer[pindex] != -1);
+                bgra pixel = palette[iBuffer[pindex]];
+
+                data[doff + 0]  = pixel.r;
+                data[doff + 1]  = pixel.g;
+                data[doff + 2]  = pixel.b;
+            }
+        }
+        else // If the distortion flag isn't set, just read the pixel data in from the palette.
+        {
+            for(std::size_t i = 0; i < header.texWidth * header.texHeight * 3; i += 3)
+            {
+                std::uint8_t index;
+                file.ReadData(&index, sizeof(index), e);
+                bgra pixel = palette[index];
+
+                data[i + 0]   = pixel.r;
+                data[i + 1]   = pixel.g;
+                data[i + 2]   = pixel.b;
+            }
+        }
+
+        Dump2Bitmap(header.texWidth, header.texHeight, data);
+    }
+    else if(header.bpp == PixelFormat::RGBA)
+    {
+        file.ReadData(&data[0], header.texSize, e);
+    }
+    else if(header.bpp == PixelFormat::BGR)
+    {
+        file.ReadData(&data[0], header.texSize, e);
+    }
+    else if(header.bpp == PixelFormat::RGBA16)
+    {
+
+        //TODO: Some kind of fucked up shit here. I think this is R5G5B5A1 or something like that..
+        file.ReadData(&data[0], header.texSize, e);
+        Log(LogLevel::WARN, "sh3_texture::Load( ): Warning: RGBA16 format not implemented!");
+    }
+    else
+    {
+        die("sh3_texture::Load( ): Unknown Pixel Format, %d", header.bpp);
+    }
+
+
+    glGenTextures(1, &tex);             // Create a texture
+    glBindTexture(GL_TEXTURE_2D, tex);  // Bind it for use
+
+    // Create the texture according to its pixel format!
+    switch(header.bpp)
+    {
+        case PixelFormat::RGBA:     // Regular 32-bit RGBA
+            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, header.texWidth, header.texHeight, 0, GL_RGBA, GL_UNSIGNED_BYTE, &data[0]);
+            break;
+        case PixelFormat::BGR:      // 24-bit BGR
+            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, header.texWidth, header.texHeight, 0, GL_BGR, GL_UNSIGNED_BYTE, &data[0]);
+            break;
+        case PixelFormat::RGBA16:   // 16-bit RGBA. OpenGL supports this (I think)
```

Comment:
It does, but differently:
`glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, ..., GL_RGBA, GL_UNSIGNED_SHORT_5_5_5_1, ...);`